### PR TITLE
Integrate Whisper model with `hg` evaluation CLI interface

### PR DESCRIPTION
**Pull request by Ahmedsaed**
**What does this PR do? Please describe:**
This PR adds integration for whisper model with the  `hg` evaluation CLI interface.

In the process, I have refactored some functions to be more generic and support huggingface transformers API.

**Demo**:
```bash
$ fairseq2 hg asr /tmp/fairseq2 --config max_num_elements=20 model_name=openai/whisper-tiny.en dtype=torch.float32
```
```
[08/09/24 19:00:49] INFO     fairseq2.recipes.hg.evaluator - Running evaluation on 1 device(s).     
eval: ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━   147   
[08/09/24 19:03:37] INFO     fairseq2.recipes.hg.evaluator - Eval Metrics - BLEU: 0.497937 | Elapsed
                             Time: 167s | Wall Time: 169s | brevity_penalty: 1.0 | length_ratio:    
                             1.1557228282673901 | precisions: [0.676166231361788,                   
                             0.5496829119972201, 0.4500613362139993, 0.3675025152851946] |          
                             reference_length: 52343 | translation_length: 60494                    
                    INFO     fairseq2.recipes.hg.evaluator - Evaluation complete in 169 seconds     
```

**Does your PR introduce any breaking changes? If yes, please list them:**
Hopefully none.

**Check list:**
- [X] Was the content of this PR **discussed and approved** via a GitHub issue? (no need for typos or documentation improvements)
- [X] Did you read the [contributor guideline](https://github.com/facebookresearch/fairseq2/blob/main/CONTRIBUTING.md)?
- [X] Did you make sure that your **PR does only one thing** instead of bundling different changes together?
- [X] Did you make sure to **update the documentation** with your changes? (if necessary)
- [X] Did you write any **new necessary tests**?
- [X] Did you verify new and **existing tests pass** locally with your changes?
- [X] Did you **update the [CHANGELOG](https://github.com/facebookresearch/fairseq2/blob/main/CHANGELOG.md)**? (no need for typos, documentation, or minor internal changes)

### CODE REVIEW

**Comment by Ahmedsaed** [comment]
I noticed that running the evaluation with `max_samples > 20` on a 15GB VRAM setup was causing a CUDA out-of-memory error. To address this, I made improvements to better manage and free up memory. With these updates, you can now run the evaluation on the entire dataset, provided the `batch_size` fits within the available memory. I successfully tested this on the whole test split (approximately 2939 examples) using `whisper-tiny` and `wav2vec2_asr_base_10h` with `max_num_elements=10`.

Additionally, I found that the text data was being encoded only to be decoded again, so I eliminated this redundant step from the pipeline, along with making other refactoring adjustments.

File: src/fairseq2/recipes/hg/asr_eval.py [hunk 1]
```diff
@@ -122,77 +150,95 @@ def _librispeech_asr_to_batch(examples: Example) -> Seq2SeqBatch:
...
+def evaluator_postprocesser(
     outputs: Any, targets: SequenceBatch, tokenizer: TextTokenizer
 ) -> tuple[list[str], list[str]]:
     decoder = tokenizer.create_decoder()
     pad_idx = tokenizer.vocab_info.pad_idx
 
     hypotheses, _ = outputs.generate_hypotheses(pad_idx=pad_idx)
     predictions = [decoder(item) for item in hypotheses]
-    references = [decoder(item) for item in targets.seqs.to(torch.int32)]
+    references = cast(list[str], targets.seqs)
```
Comment (antoine-tran) [inline src/fairseq2/recipes/hg/asr_eval.py:200]: Are you sure all targets.seqs will have torch.int32 ?

File: src/fairseq2/recipes/hg/asr_eval.py [hunk 2]
```diff
@@ -122,77 +150,95 @@ def _librispeech_asr_to_batch(examples: Example) -> Seq2SeqBatch:
...
+    """
+    try:
+        retrieve_asset_card(config.model_name)
+        return load_wav2vec2_asr_evaluator(config, output_dir)
+    except AssetNotFoundError:
+        return load_hg_asr_evaluator(config, output_dir)
+
+
 def load_wav2vec2_asr_evaluator(
     config: AsrEvalConfig, output_dir: Path
```
Comment (antoine-tran) [inline src/fairseq2/recipes/hg/asr_eval.py:226]: We can take this chance to put `output_dir` as part of AsrEvalConfig too

File: src/fairseq2/recipes/hg/asr_eval.py [hunk 3]
```diff
@@ -232,6 +271,79 @@ def load_wav2vec2_asr_evaluator(
         gang=gang,
         data_reader=pipeline_reader,
         wall_watch=wall_watch,
-        preprocessor=seq2seq_preprocessor,
-        postprocessor=lambda x, y: postprocesser(x, y, tokenizer),
+        preprocessor=evaluator_preprocessor,
+        postprocessor=lambda x, y: evaluator_postprocesser(x, y, tokenizer),
```
Comment (antoine-tran) [inline src/fairseq2/recipes/hg/asr_eval.py:275]: let's use `functools.partial` for better readability

File: src/fairseq2/recipes/hg/asr_eval.py [hunk 4]
```diff
@@ -232,6 +271,79 @@ def load_wav2vec2_asr_evaluator(
...
+    )
+
+
+class HGModelWrapper:
+    def __init__(self, model: WhisperForConditionalGeneration):
```
Comment (antoine-tran) [inline src/fairseq2/recipes/hg/asr_eval.py:280]: Why do we have to tie to `WhisperForConditionalGenereation` here ?

File: src/fairseq2/recipes/hg/asr_eval.py [hunk 5]
```diff
@@ -232,6 +271,79 @@ def load_wav2vec2_asr_evaluator(
...
         data_reader=pipeline_reader,
         wall_watch=wall_watch,
-        preprocessor=seq2seq_preprocessor,
-        postprocessor=lambda x, y: postprocesser(x, y, tokenizer),
+        preprocessor=evaluator_preprocessor,
+        postprocessor=lambda x, y: evaluator_postprocesser(x, y, tokenizer),
+    )
+
+
+class HGModelWrapper:
```
Comment (antoine-tran) [inline src/fairseq2/recipes/hg/asr_eval.py:279]: Thought: A better way to handle this is to have a HG model loader with HG model config, where we can specify the transformers class and the preprocessor class. But this can be left to another PR

File: src/fairseq2/recipes/hg/asr_eval.py [hunk 6]
```diff
@@ -232,6 +271,79 @@ def load_wav2vec2_asr_evaluator(
...
+    wall_watch = Stopwatch(start=True, device=init_device)
+
+    return HFEvaluator[Seq2SeqBatch](
+        model=cast(Model, HGModelWrapper(model)),
+        metrics=["bleu"],
+        gang=gang,
+        data_reader=pipeline_reader,
+        wall_watch=wall_watch,
+        preprocessor=evaluator_preprocessor,
+        postprocessor=lambda x, y: (
```
Comment (antoine-tran) [inline src/fairseq2/recipes/hg/asr_eval.py:345]: let's put this in a function

File: src/fairseq2/recipes/hg/asr_eval.py [hunk 7]
```diff
@@ -232,6 +271,79 @@ def load_wav2vec2_asr_evaluator(
...
+
+    processor = WhisperProcessor.from_pretrained(config.model_name)
+    model = WhisperForConditionalGeneration.from_pretrained(config.model_name).to(
+        init_device
+    )
+
+    ds_builder = load_dataset_builder(config.dataset_name)
+    ds = prepare_dataset(
+        config,
+        lambda x: {
```
Comment (antoine-tran) [inline src/fairseq2/recipes/hg/asr_eval.py:318]: let's put this in a separate function to avoid to nested parameters

File: src/fairseq2/recipes/hg/asr_eval.py [hunk 8]
```diff
@@ -232,6 +271,79 @@ def load_wav2vec2_asr_evaluator(
...
+        raise ValueError(f"Expect AsrEvalConfig, get {type(config)}")
+
+    gang = setup_root_gang(log)
+
+    if gang.rank == 0:
+        init_device = gang.device
+    else:
+        init_device = META
+
+    processor = WhisperProcessor.from_pretrained(config.model_name)
```
Comment (antoine-tran) [inline src/fairseq2/recipes/hg/asr_eval.py:310]: We can make this dynamic by specify the class of model and processor, then use the library hydra to load them:

model = hydra.utils.get_class(model_class)
processor = hydra.utils.get_class(processor)

As similar to `datasets`, `evaluate` and `transformers`, the import of `hydra-core` should come with a warning if not installed before.

File: src/fairseq2/recipes/hg/asr_eval.py [hunk 9]
```diff
@@ -232,6 +271,79 @@ def load_wav2vec2_asr_evaluator(
...
+        converter=lambda x: to_batch(x, "whisper", init_device),
+        batching=StaticBatching(config.max_num_elements),
+        num_prefetch=config.num_prefetch,
+        max_seq_len=config.max_audio_len,
+    )
+
+    wall_watch = Stopwatch(start=True, device=init_device)
+
+    return HFEvaluator[Seq2SeqBatch](
+        model=cast(Model, HGModelWrapper(model)),
```
Comment (antoine-tran) [inline src/fairseq2/recipes/hg/asr_eval.py:339]: See my other comment. 

This forced casting is artificial IMHO - The better option is to create a [model loader](https://github.com/facebookresearch/fairseq2/blob/main/src/fairseq2/models/loader.py#L70) that creates a HG model out of its name. But let's make it in a separate PR.

**Review by antoine-tran — COMMENTED** [review]
Thanks for this PR !

I can see you have made attempts to remove duplicates from the first iteration and parameterize some functions such as `prepare_dataset()`, `load_asr_evaluator()`, etc.

I think we need a few more building blocks to make this PR a solid one (dynamic loading of HG model and processor, improve the `AsrEvalConfig` to cover `output_dir` , better handling of inline functions). There should be a separate PR to handle the HG model loader, too, but let's discuss this point in our call

File: src/fairseq2/recipes/hg/asr_eval.py (hunk 1 above)
Comment (Ahmedsaed) [reply src/fairseq2/recipes/hg/asr_eval.py:200]: At this point target seqs are just the strings themselves.

I have removed the encoding and decoding of text since its redundant (text is being encoded only for it to be decoded again without using the tokens)

**Comment by Ahmedsaed** [comment]
I will close this PR in favor of 2 smaller PRs
- `AsrDatasetConfig` which will also include all the refactoring made in this PR along with a config for handling ASR Datasets.
- `Whisper Integration` will just include the whisper related code.

File: src/fairseq2/recipes/hg/asr_eval.py (hunk 2 above)
Comment (Ahmedsaed) [reply src/fairseq2/recipes/hg/asr_eval.py:226]: As discussed, This is part of the `RecipeCommandHandler` so I can't modify it without affecting the whole project.

I will mark this comment as resolved.

File: src/fairseq2/recipes/hg/asr_eval.py (hunk 3 above)
Comment (Ahmedsaed) [reply src/fairseq2/recipes/hg/asr_eval.py:275]: I have done this in both `AsrDatasetConfig` and `whisper` PRs.

File: src/fairseq2/recipes/hg/asr_eval.py (hunk 4 above)
Comment (Ahmedsaed) [reply src/fairseq2/recipes/hg/asr_eval.py:280]: I have addressed this in the new whisper PR by using `PreTrainedModel` instead.

File: src/fairseq2/recipes/hg/asr_eval.py (hunk 7 above)
Comment (Ahmedsaed) [reply src/fairseq2/recipes/hg/asr_eval.py:318]: I have addressed this in the new whisper PR

File: src/fairseq2/recipes/hg/asr_eval.py (hunk 8 above)
Comment (Ahmedsaed) [reply src/fairseq2/recipes/hg/asr_eval.py:310]: I have done this in PR https://github.com/facebookresearch/fairseq2/pull/751

**Comment by Ahmedsaed** [comment]
I have addressed all comments in the following PRs #749 and #751
